app/scripts/migrations/148.ts

```diff
@@ -0,0 +1,87 @@
+import { cloneDeep } from 'lodash';
+import { hasProperty, isObject } from '@metamask/utils';
+import {
+  Caip25EndowmentPermissionName,
+  Caip25CaveatType,
+} from '@metamask/chain-agnostic-permission';
+
+export const version = 148;
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+/**
+ * Add sessionProperties property to CAIP-25 permission caveats
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state
+ * @returns Updated versioned MetaMask extension state
+ */
+export async function migrate(originalVersionedData: VersionedData) {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  transformState(versionedData.data);
+  return versionedData;
+}
+
+/**
+ * Add sessionProperties property to CAIP-25 permission caveats
+ *
+ * @param state - The MetaMask state
+ * @returns The updated MetaMask state
+ */
+function transformState(state: VersionedData['data']) {
+  if (!hasProperty(state, 'PermissionController')) {
+    return state;
+  }
+
+  if (!isObject(state.PermissionController)) {
+    global.sentry?.captureException?.(
+      new Error(
+        `Migration ${version}: typeof state.PermissionController is ${typeof state.PermissionController}`,
+      ),
+    );
+    return state;
+  }
+
+  const { subjects } = state.PermissionController;
+
+  if (!isObject(subjects)) {
+    global.sentry?.captureException?.(
+      new Error(
+        `Migration ${version}: typeof state.PermissionController.subjects is ${typeof subjects}`,
+      ),
+    );
+    return state;
+  }
+
+  for (const subject of Object.values(subjects)) {
+    if (
+      !isObject(subject) ||
+      !hasProperty(subject, 'permissions') ||
+      !isObject(subject.permissions)
+    ) {
+      continue;
+    }
+
+    const { permissions } = subject;
+    const caip25Permission = permissions[Caip25EndowmentPermissionName];
+
+    if (
+      !isObject(caip25Permission) ||
+      !Array.isArray(caip25Permission.caveats)
+    ) {
+      continue;
+    }
+
+    for (const caveat of caip25Permission.caveats) {
+      if (caveat.type === Caip25CaveatType && isObject(caveat.value)) {
+        if (!hasProperty(caveat.value, 'sessionProperties')) {
+          caveat.value.sessionProperties = {};
+        }
+      }
```

Comment:
okay if we add a `break` in here? or use `find` instead of the for loop? definitely a nit

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/30949/commits/2ea1fcdcee5b15741ceaa19030caf1a8290a1c31 & here https://github.com/MetaMask/metamask-extension/pull/30949/commits/4450c7e31d3859cb4d2b0ab9964640008a38a655